### ISSUE TITLE
[Issue] Energy Detector sometimes voids energy, or is an infinit energy source without input

### ISSUE THREAD

**Issue by bessw** [Seniorendi/AdvancedPeripherals#52]
# Describe

## Expected

- measures the throughput and allows to limit it
- input and output of the Energy Detector are always the same
- no energy is lost (or only a bit as energy consumption of the Energy Detector)

## Observed

1. one time it suddenly was an infinity energy source even without an input attached (it gave 1.6 kFE/t)
2. now it drains any available energy on the input and voids the excess
    - multiple in parallel share the input evenly
    - limiting the throughput works, but it is used up to max even without an attached cable
3. two of our Energy Detectors are limited to 0 and don't have an input source attached, but still report a throughput of last time it was transmitting power

# Reproduce

1. use an not infinite energy source (Bigreactors - Basic Reactor)

## used mods
- Bigreactors - Basic Reactor
- Flux Networks
- Rftoolspower - Powercell (High)
- Makanism - Basic Universal Cable
- many others not directly interfacing with this (ask if you need to know more)

**Versions:**

- Forge version: 36.1.4
- AdvancedPeripherals version: 0.4.4b

**Comment by undercoverturbo** [comment]
FYI, Rebooting the PC did not fix the report problem of the third observed behaviour.

**Comment by bessw** [comment]
By looking at [EnergyDetectorTileEntity.java](https://github.com/Seniorendi/AdvancedPeripherals/blob/3d12e8805176925bd18181082dd13168c86e74d2/src/main/java/de/srendi/advancedperipherals/common/blocks/tileentity/EnergyDetectorTileEntity.java)
it looks to me like the internal input and output of the Energy Detector is in fact a two times int_max sized energy storage, thus what we thought is voided is instead stored in the Energy Detector and is later what we thought is an infinity energy source.

**Comment by Seniorendi** [comment]
When you use setTransferRateLimit([EnergyDetectorPeripheral.java](https://github.com/Seniorendi/AdvancedPeripherals/blob/3d12e8805176925bd18181082dd13168c86e74d2/src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/EnergyDetectorPeripheral.java)) these values will override. But yes, that is the issue. I will try to fix that today if I can find time.

**Comment by Seniorendi** [comment]
Oops, i wouldn't close the issue.

**Comment by Seniorendi** [comment]
Hey, could you try if this version fixed your issue?
[advancedperipherals-0.4.5-DEVb.zip](https://github.com/Seniorendi/AdvancedPeripherals/files/6324471/advancedperipherals-0.4.5-DEVb.zip)

**Comment by bessw** [comment]
No, its still the same.

**Comment by Seniorendi** [comment]
[advancedperipherals-0.4.5-DEV1b.zip](https://github.com/Seniorendi/AdvancedPeripherals/files/6326561/advancedperipherals-0.4.5-DEV1b.zip)
I think, I fixed it. In any case, I can no longer reproduce it. 
Unzip the file before you use it.

**Comment by Seniorendi** [comment]
Ahh yes, I can reproduce this from you above too. The energy detector eats the energy because the storage is quite infinite big.
I fixed that.
[advancedperipherals-0.4.5-DEV2b.zip](https://github.com/Seniorendi/AdvancedPeripherals/files/6326728/advancedperipherals-0.4.5-DEV2b.zip)
There is now also a config option for the maximum flow range.
You can test it if you want.
Unzip the file before you use it.

**Comment by bessw** [comment]
While im testing just a thought:

I don't know the Energy Transfer Api, but is there a specific reason why you need an internal storage on the Energy Detector?
Is it maybe possible to directly transfer the energy from the source block to the output block?

**Comment by Seniorendi** [comment]
Well, this is a very good idea.
I will try that out when i am at home again.

**Comment by bessw** [comment]
> Ahh yes, I can reproduce this from you above too. The energy detector eats the energy because the storage is quite infinite big.
> I fixed that.
> [advancedperipherals-0.4.5-DEV2b.zip](https://github.com/Seniorendi/AdvancedPeripherals/files/6326728/advancedperipherals-0.4.5-DEV2b.zip)
> There is now also a config option for the maximum flow range.
> You can test it if you want.
> Unzip the file before you use it.

Now it doesn't transfer any energy. (I also tried it in a new world)

**Comment by bessw** [comment]
Wait, now in an other setup it works, I will need to take a closer look on what was wrong there.

**Comment by Seniorendi** [comment]
I changed the system of the energy detector completely. It will send the energy from the tile entity of the one side to the tile entity of the output side. It will no longer store any energy.

**Comment by bessw** [comment]
Ok great now its transferring energy and it doesn't store energy, but sadly there are new problems:

The Energy Detector is now actively draining energy out of the cable (instead of passive receiving it), this has the consequence that if there isn't enough energy for all devices that need energy the Energy Detector takes as much as it needs and all other devices only get the rest (if there is any).

This behavior is different to how cables normally work, because normally they would evenly distribute the available energy over the devices.

![image](https://user-images.githubusercontent.com/56679598/115127243-5f42dd00-9fd5-11eb-8000-f49e37c87fbe.png)


## idea for solution:
You could implement a proxy class for the energy storage on the output side and set it as the "internal storage" of the Energy Detector:
- when ever the cable tries to put energy into the Energy Detector the EnergyStorageProxy looks if there is an output device and forwards the the method calls to it
- if there is no output device it should return that there is no storage capacity left / maybe also canReceive() -> false
- `EnergyStorageProxy.receiveEnergy(inEnergy, ...)` should be something like
    ```java
    int transferred= out.receiveEnergy(Math.min(inEnergy, maxTransferRate), ...)
    setTransferrate(transferred)
    return transferred
    ```

Btw. the current transferrate returned by the peripheral shows sometimes old data.
Maybe you should sum `transferred` over one tick and check on every tick if you need to update the peripheral, because receiveEnergy() could be called multiple times.

**Comment by Seniorendi** [comment]
> The Energy Detector is now actively draining energy out of the cable (instead of passive receiving it)

I fixed that, that also fixed that the peripheral shows old data.

> EnergyStorageProxy.receiveEnergy(inEnergy, ...) should be something like

I already set the transfer rate to the received energy.

> This behavior is different to how cables normally work, because normally they would evenly distribute the available energy over the devices.

Mekanism can't send the energy if the energy detector use it. 
If you decrease the max transfer rate of the second energy detector, the universal cable will be able to send energy to the other cable.

**Comment by bessw** [comment]
> I fixed that, that also fixed that the peripheral shows old data. 👍🏻

Great, I will take a look at it once the update is out.

As far as I understand your code in your latest commit 12f1c52 you just moved the "if no input or no output return null" to the top of the method, or did I miss something?
How does this fix that the energy detector is actively draining energy from the input?

> I already set the transfer rate to the received energy.

I just wrote `setTransferrate(transferred)` because how you update the transferrate is not relevant for what I wanted to suggest, maybe that was more confusing then helpful.

Actually I was thinking about how the `receiveEnergy(int maxReceive, boolean simulate)` Method of a `EnergyStorageProxy` class could look like to handle passively receiving energy, this `EnergyStorageProxy` would then be set as the internal storage of the energy detector.

> Mekanism can't send the energy if the energy detector use it.
> If you decrease the max transfer rate of the second energy detector, the universal cable will be able to send energy to the other cable.

My point was that it behaves differently if I replace the second energy detector with a universal cable and that my expectation was that a energy detector would behave the same like a universal cable as long as both don't exceed their max transfer rate.
(at least in this setup, I know that the energy detector behaves as a diode, but shouldn't make a difference in this setup)

**Comment by Seniorendi** [comment]
> you just moved the "if no input or no output return null" to the top of the method, or did I miss something?
I used && instead of || before, so the energy detector will drain energy even if there is no tile entity.

> I just wrote setTransferrate(transferred) because how you update the transferrate is not relevant for what I wanted to suggest, maybe that was more confusing then helpful.
I now know what you mean, sorry. I'll look into it tomorrow.

> My point was that it behaves differently if I replace the second energy detector with a universal cable and that my expectation was that a energy detector would behave the same like a universal cable as long as both don't exceed their max transfer rate.

Sorry, but I do not know how I could fix that.  I'll look into it tomorrow, but I can't guarantee anything

**Comment by bessw** [comment]
I just managed to build your mod on my pc, I will also try to take a look into it.

**Comment by bessw** [comment]
I think I fixed the transferring so that it works as I expected,
but now I'm still searching the reason why the current transferrate returned by the peripheral shows sometimes old data.

I'll look later into it again and create a pull request after some more experiments.

**Comment by Seniorendi** [comment]
> I'll look later into it again and create a pull request after some more experiments.

Thank you for your help 👍🏻

**Comment by Seniorendi** [comment]
> why the current transferrate returned by the peripheral shows sometimes old data.
I can't reproduce that, the detector returns always the current data

**Comment by Seniorendi** [comment]
I should review before I merge a pull request...
The energy detector do not send energy out/do not receive energy.
I hadn't much time to look into this, but could you try to fix that?

**Comment by bessw** [comment]
That's strange...
I just checked out your master branch after the merge and tested it locally, for me it works without any problems.

Could you please tell me what you did / how you hooked it up, so I can try to reproduce it?

**Comment by Seniorendi** [comment]
![image](https://user-images.githubusercontent.com/67484093/115423796-68fb5900-a1fe-11eb-962d-3bb9cad94ff5.png)

![image](https://user-images.githubusercontent.com/67484093/115423867-76b0de80-a1fe-11eb-8996-ac4c6d8ac174.png)

It does not seem that you send the energy? Or do I misunderstand anything?

**Comment by bessw** [comment]
It does send energy, but it still seams like there is strange stuff happening with universal cables.


using the Network Reader at the input cable:
![input cable](https://user-images.githubusercontent.com/56679598/115432575-d27f6580-a206-11eb-9460-ffe48a9dc4f4.png)

![computer](https://user-images.githubusercontent.com/56679598/115432291-80d6db00-a206-11eb-84dc-29595c2960f9.png)

using the Network Reader at the output cable:
![output cable](https://user-images.githubusercontent.com/56679598/115432508-bb407800-a206-11eb-991f-5aa33f279b26.png)

![storage](https://user-images.githubusercontent.com/56679598/115432129-5422c380-a206-11eb-9f19-151c27070043.png)

- The input cable seems to call `receiveEnergy(...)` 4 times per tick => yes my code still makes it possible that the maxTransferrate can be bypassed by using a cable that calls `receiveEnergy(...)` multiple times per tick. The only difference to your code is that mine prints the sum of the transferred energy of all calls to `receiveEnergy(...)` in the last tick and yours only prints the transferred energy from the last call to `receiveEnergy(...)`.

- What also confuses me is that the cable after the energy detector has a throughput of 9.5FE/t but the induction Matrix only receives 5FE/t like the actual limit and the input cable was.

Could it be something because `getCapability(...)` returns for both sides the same storage?

**Comment by bessw** [comment]
But I still don't see any reason why yours doesn't transfer energy...

**Comment by Seniorendi** [comment]
I will try to push the pull request and see if that works for me.

**Comment by Seniorendi** [comment]
There is an issue with your commit.
You need to use `Optional#isPresent` instead of `isEmpty`. `isPresent `would return true if the Optional isn't empty. So you should use `!isEmpty`.

The issue is at EnergyDetectorTileEntity.java:103. I could fix that, but i'm busy. Sorry

**Comment by bessw** [comment]
The if block doesn't need to be entered if outReceivingStorage is present, because it is only there to refetch the capability if the lazyOptional has been invalidated.
This is just a performance optimization because the documentation says:
> Modders are encouraged to cache this value, using the listener capabilities of the Optional to be notified if the requested capability get lost.

**Comment by bessw** [comment]
![image](https://user-images.githubusercontent.com/56679598/115450289-fef1ac80-a21b-11eb-9506-56c5d74f6011.png)

oops that's a java 11 feature, sorry

**Comment by Seniorendi** [comment]
Yeah, I need to use java 8, sorry.

**Comment by Seniorendi** [comment]
Stupid auto close... 😄

**Comment by Seniorendi** [comment]
![image](https://user-images.githubusercontent.com/67484093/115451908-0e71f500-a21e-11eb-851e-4e01f8cb1519.png)

Well, thank you for your help! I really appreciate that!

**Comment by Seniorendi** [comment]
Fixed in 0.4.6b
